hgeometry/src/Algorithms/Geometry/MonotonePolygon/MonotonePolygon.hs

```diff
@@ -64,42 +66,61 @@ isMonotone direction p = all isMonotoneAt (map _core $ toPoints p)
 randomMonotone nVertices direction = do
     -- 1, skip 2 in this function bc `direction` is given
-    points <- Data.List.replicateM createRandomPoint nVertices
+    points <- replicateM nVertices createRandomPoint
     -- 3
     let 
-        min = Data.Geometry.Polygon.Core.minimumBy (cmpExtreme direction) points
-        max = Data.Geometry.Polygon.Core.maximumBy (cmpExtreme direction) points
+        specialPoints = map (\x -> x :+ ()) points
+        min = Data.List.minimumBy (cmpExtreme direction) specialPoints
+        max = Data.List.maximumBy (cmpExtreme direction) specialPoints
         -- 4
-        pointsWithoutExtremes = filter (\x -> x /= min && x /= max) points
+        pointsWithoutExtremes = filter (\x -> x /= min && x /= max) specialPoints
         -- 5, 6
-        (leftHalfUnsorted,rightHalfUnsorted) = Data.List.partition (\x -> ccw min max x == CCW) pointsWithoutExtremes
+        (leftHalfUnsorted,rightHalfUnsorted) = Data.List.partition (toTheLeft min max) pointsWithoutExtremes
         leftHalf = sortBy (cmpExtreme direction) leftHalfUnsorted
-        rightHalf = sortBy (cmpExtreme (reverse direction)) rightHalfUnsorted
+        rightHalf = reverse (sortBy (cmpExtreme direction) rightHalfUnsorted)
     return (fromPoints ([min] ++ leftHalf ++ [max] ++ rightHalf))
 
 -- Pick a random vector and then call 'randomMonotone'.
 -- | \( O(n \log n) \)
 randomMonotoneDirected :: RandomGen g => Int -> Rand g (SimplePolygon () Rational)
-randomMonotoneDirected nVertices = randomMonotone nVertices direction
-    where
-        direction = case (Data.Geometry.Vector.vectorFromList createRandomRationalVec2) of
-                        Just a -> a
-                        Nothing -> error "could not create random vector"
+randomMonotoneDirected nVertices = do
+    points <- replicateM nVertices createRandomPoint
+    direction <- generateRandomVector2
+    let 
+        specialPoints = map (\x -> x :+ ()) points
+        min = Data.List.minimumBy (cmpExtreme direction) specialPoints
+        max = Data.List.maximumBy (cmpExtreme direction) specialPoints
+        -- 4
+        pointsWithoutExtremes = filter (\x -> x /= min && x /= max) specialPoints
+        -- 5, 6
+        (leftHalfUnsorted,rightHalfUnsorted) = Data.List.partition (toTheLeft min max) pointsWithoutExtremes
+        leftHalf = sortBy (cmpExtreme direction) leftHalfUnsorted
+        rightHalf = reverse (sortBy (cmpExtreme direction) rightHalfUnsorted)
+    return (fromPoints ([min] ++ leftHalf ++ [max] ++ rightHalf))
 
 -------------------------------------------------------------------------------------------------
 -- helper functions
 
 granularity :: Integer
 granularity = 10000000
 
+toTheLeft :: Point 2 Rational :+ () -> Point 2 Rational :+ () -> Point 2 Rational :+ () -> Bool
+toTheLeft min max x = Data.Geometry.Point.Orientation.Degenerate.ccw a b c == CCW
+    where
+        a = _core min
+        b = _core max
+        c = _core x

```

Comment:
The `ccw'` helper function calls `_core` for you.

`ccw' :: (Ord r, Num r) => Point 2 r :+ a -> Point 2 r :+ b -> Point 2 r :+ c -> CCW`